templates/common/alibabacloud/files/usr-local-lib-systemd-system-generators-alibaba-kubelet-nodename-generator.sh.yaml

```diff
@@ -0,0 +1,36 @@
+mode: 0755
+path: "/usr/local/lib/systemd/system-generators/alibaba-kubelet-nodename-generator.sh"
+contents:
+  inline: |
+    #!/bin/sh
+    set -e -o pipefail
+
+    # This systemd.generator(7) determines the alibaba nodename and sets
+    # it as a variable for the kubelet.service unit file.
+    #
+    # Place in /usr/local/lib/systemd/system-generators
+    #
+
+
+    # If invoked with no arguments (for testing) write to /tmp.
+    earlydir="/tmp"
+    if [ -n "$2" ]; then
+        earlydir="$2"
+    fi
+
+    kubelet_service_d_dir="$earlydir/kubelet.service.d"
+    node_conf="$kubelet_service_d_dir/alibaba-node-name.conf"
+
+    # Set KUBE_NODE_NAME to be overridden by --hostname-override in the kubelet.
+    # Format required is "<region_id>.<instance_id>"
+    # https://github.com/kubernetes/cloud-provider-alibaba-cloud/blob/master/docs/getting-started.md
+    meta_ep=http://100.100.100.200/latest/meta-data
+    region_id=$(curl -s ${meta_ep}/region-id)
+    instance_id=$(curl -s ${meta_ep}/instance-id)
```

Comment:
Hmm actually this is tricky because systemd generators run *really* early and it's possible we won't have networking up then.  So we'd want to retry.

I think where this is going actually is to append to an environment file...in fact, we already have a pile of these:

```
  EnvironmentFile=-/etc/kubernetes/kubelet-workaround
  EnvironmentFile=-/etc/kubernetes/kubelet-env
  EnvironmentFile=/etc/node-sizing.env
```

This file will be read only when systemd goes to start kubelet - this way we can do what we need to as a unit that runs `Before=kubelet.service` as before.

To say it another way, by writing to an `EnvironmentFile` instead of mutating `Environment=` we don't need to reload systemd.

Reply:
To do this slightly more elegantly, you could have the generator synthesize the unit fragment

```
[Service]
EnvironmentFile=/run/kubelet-alibaba.env
```

Then the systemd unit would populate that file.

Notice that we don't have `-` here - we would fail if it doesn't exist.